pkg/geo/geomfn/subdivide.go

```diff
@@ -0,0 +1,364 @@
+		return errors.Newf("error checking if geom.T is collection: %v", err)
+	}
+	_, isMultiPoint := gt.(*geom.MultiPoint)
+	if isCollection && !isMultiPoint {
+		err := forEachGeomTInCollection(gt, func(gi geom.T) error {
+			// It's not a subdivision yet, so depth stays the same
+			return subdivideRecursive(gi, maxVertices, depth, col, dim)
+		})
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	currDim, err := dimensionFromGeomT(gt)
+	if err != nil {
+		return errors.Newf("error checking geom.T dimension: %v", err)
+	}
+	if currDim < dim {
+		// Ignore lower dimension object produced by clipping at a shallower recursion level.
+		return nil
+	}
+	if depth > maxDepth {
+		err = col.Push(gt)
+		if err != nil {
+			return errors.Newf(geometryCollectionPushError, err)
+		}
+		return nil
+	}
+	nVertices, err := countVertices(gt)
+	if err != nil {
+		return errors.Newf("error counting vertices: %v", err)
+	}
+	if nVertices == 0 {
+		return nil
+	}
+	if nVertices <= maxVertices {
+		err = col.Push(gt)
+		if err != nil {
+			return errors.Newf(geometryCollectionPushError, err)
+		}
+		return nil
+	}
+
+	splitHorizontally := width > height
+	splitPoint, err := calculateSplitPoint(gt, nVertices, splitHorizontally, clip)
+	if err != nil {
+		return err
+	}
+	subBox1 := *clip
+	subBox2 := *clip
+	if splitHorizontally {
+		subBox1.HiX = splitPoint
+		subBox2.LoX = splitPoint
+	} else {
+		subBox1.HiY = splitPoint
+		subBox2.LoY = splitPoint
+	}
+
+	clipped1, err := clipGeomTByBoundingBox(gt, &subBox1)
+	if err != nil {
+		return errors.Newf("error clipping geom.T: %v", err)
+	}
+	clipped2, err := clipGeomTByBoundingBox(gt, &subBox2)
+	if err != nil {
+		return errors.Newf("error clipping geom.T: %v", err)
+	}
+	depth++
+	err = subdivideRecursive(clipped1, maxVertices, depth, col, dim)
+	if err != nil {
+		return err
+	}
+	err = subdivideRecursive(clipped2, maxVertices, depth, col, dim)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// isGeomTCollection returns whether the given geom.T is of a collection type (wrapper for a geometry function).
+func isGeomTCollection(gt geom.T) (bool, error) {
+	g, err := geo.MakeGeometryFromGeomT(gt)
+	if err != nil {
+		return false, err
+	}
+	return IsCollection(g)
+}
+
+// forEachGeomTInCollection calls the provided function for each geometry in the passed in collection.
+// Ignores MultiPoint input.
+func forEachGeomTInCollection(gt geom.T, fn func(geom.T) error) error {
+	switch gt := gt.(type) {
+	case *geom.GeometryCollection:
+		for i := 0; i < gt.NumGeoms(); i++ {
+			err := fn(gt.Geom(i))
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	case *geom.MultiPolygon:
+		for i := 0; i < gt.NumPolygons(); i++ {
+			err := fn(gt.Polygon(i))
+			if err != nil {
+				return err
+			}
+		}
+	case *geom.MultiLineString:
+		for i := 0; i < gt.NumLineStrings(); i++ {
+			err := fn(gt.LineString(i))
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func calculateSplitPoint(
```

Comment:
nit: `calculateSplitPointCoordForSubdivide`, also a comment

Reply:
I'd also just pass a copy of clip into here.

Reply:
Makes sense, fixed